File: src/lib/tool-manifest/tools/highlight-region.ts
import { z } from 'zod';
import { useEditorStore } from '@/store';
import { useAiSession } from '@/hooks/useImageContext';
import { ackSchema } from '../shared-schemas';
import type { ToolManifest } from '../types';

const input = z.object({
  label: z.string().describe('Region label as returned by list_named_regions.'),
  reasoning: z.string().optional().describe('Why are you pointing at this region? (Shown in the future annotation UI.)'),
});

export const highlightRegionTool: ToolManifest<typeof input, typeof ackSchema> = {
  name: 'highlight_region',
  kind: 'emit',
  description:
    'Visually point at a region for the user without committing it as a selection-for-adjustment. Use this when you want the user to *look* at something — not when you want to act on it.',
  usage:
    'Currently implemented as a transient active-mask preview using the Plan 1 overlay substrate. Future work: dedicated annotation overlay layer with the reasoning text rendered alongside.',
  inputSchema: input,
  outputSchema: ackSchema,
  handler: ({ label, reasoning }) => {
    const ctx = useAiSession.getState().context;
    const region = ctx?.candidateRegions?.find((r) => r.label === label);
    if (!region?.maskRef) {
      return { ok: false, message: `No region with label "${label}" has a mask.` };
    }
    // Arm as active (preview) only — don't commit. The overlay substrate will
    // render fill + marching-ants outline immediately.
    useEditorStore.getState().setActiveMask(region.maskRef);
    return {
      ok: true,
      message: reasoning ? `Highlighting "${label}": ${reasoning}` : `Highlighting "${label}".`,
    };
  },
};

File: backend/app/tools/atomic/highlight_region.py
from __future__ import annotations

from pydantic import BaseModel, Field

from app.state.document import SessionDocument
from app.tools.atomic.select_named_region import _ScopeUnresolvable, _UnknownRegion
from app.tools.base import BackendTool, ToolPermissions


class _Input(BaseModel):
    label: str = Field(min_length=1)
    reasoning: str | None = None


class _Output(BaseModel):
    ok: bool
    mask_id: str


class HighlightRegionTool(BackendTool[_Input, _Output]):
    name = "highlight_region"
    kind = "emit"
    description = (
        "Visually point at a region for the user without committing it as a selection. "
        "Use this to draw attention; not to act on."
    )
    input_schema = _Input
    output_schema = _Output
    permissions = ToolPermissions(requires_image=False)

    async def handler(self, doc: SessionDocument, input: _Input) -> _Output:  # noqa: A002
        ctx = doc.image_context
        if ctx is None:
            raise _ScopeUnresolvable("no image context yet")
        if not any(r.label == input.label for r in ctx.candidate_regions):
            raise _UnknownRegion(input.label)
        mask = next((m for m in doc.masks.values() if m.label == input.label), None)
        if mask is None:
            raise _ScopeUnresolvable(f"region {input.label!r} has no registered mask")
        doc.active_mask_id = mask.id
        doc.emit_selection_changed(mask.id, "active", input.label)
        return _Output(ok=True, mask_id=mask.id)
